Implement promise bindings

jdm opened the pull request:
This implements support for using Promises in WebIDL, executing promise callbacks in batches (equivalent to running all enqueued jobs from a setTimeout(0)), and attaching native callbacks to promise objects. This is the combined work of myself, @dati91, and @mmatyas based on the following prior work in Gecko:
- Codegen.py
- Promise.webidl
- Promise.cpp/Promise.h
- PromiseNativeHandler.h

This does not implement microtasks per #4283. This allows us to make progress on testing code that requires the use of Promises right now; the microtasks work is more complicated, but also largely orthogonal to implement.

Requires https://github.com/servo/mozjs/pull/89, https://github.com/servo/rust-mozjs/pull/287, and https://github.com/servo/rust-mozjs/pull/294.

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #4282
- [X] There are tests for these changes

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12830)

<!-- Reviewable:end -->

jdm committed:
Implement binding support for returning and accepting Promises in WebIDL.

mmatyas committed:
Enqueue promise jobs from SpiderMonkey callbacks, and execute them in batches. Implement native Promise APIs.
Add SpiderMonkey hooks for enqueuing promise jobs. Start porting various native Promise APIs.

jdm committed:
Useful APIs for resolving/rejecting from native code.

jdm committed:
Support native promise callbacks.

jdm committed:
Add integration tests for interacting with promises from native code.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/bindings/codegen/CodegenRust.py, components/script/dom/bindings/global.rs, components/script/dom/webidls/PromiseNativeHandler.webidl, components/script/dom/testbinding.rs, components/script/dom/mod.rs, components/script/dom/bindings/codegen/Bindings.conf, components/script/dom/promisenativehandler.rs, components/script/dom/webidls/TestBinding.webidl, components/script/script_thread.rs, components/script/dom/bindings/js.rs, components/script/dom/webidls/Promise.webidl, components/script/dom/promise.rs, components/script/dom/bindings/codegen/Configuration.py

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits include an empty title element (`<title></title>`). Consider adding appropriate metadata.
* These commits modify **unsafe code**. Please review it carefully!

Wafflespeanut commented:
r? @Ms2ger, @nox, and everyone else :smile:

jdm commented:
This PR does not support workers yet. That should only require moving the various queues out of ScriptThread and into Runtime instead.

jdm commented:
cc @creativcoder ;)

nox commented:
@jdm It doesn't compile.

nox commented on components/script/dom/bindings/global.rs:
```diff
@@ -287,6 +288,19 @@ impl<'a> GlobalRef<'a> {
+    fn init_reflector(&mut self, _obj: *mut JSObject) {
+        unreachable!()
+    }
+}
```
Why is this impl needed?

jdm replied:
So we can pass a GlobalRef to a function that requires a T: Reflectable.

jdm commented:
@nox It doesn't compile because it relies on open PRs in other repositories.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #12970) made this pull request unmergeable. Please resolve the merge conflicts.

nox commented:
Didn't see anything bad and just have a few questions. @Ms2ger Would be great if you could skim through it quickly, given you know SM better than me.

-S-awaiting-review +S-needs-squash 

---

Reviewed 10 of 10 files at r1, 4 of 4 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 4 of 4 files at r5, 6 of 6 files at r6, 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[components/script/dom/bindings/global.rs, line 302 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12830#-KOzs7Gn2sILEkC6p9pH-r7-302:-KPrteBU4NvYlxMXuFcR:b-d56cb1) ([raw file](https://github.com/servo/servo/blob/c5ee871ce9a0bd5b56fd4e66ce992590c749935e/components/script/dom/bindings/global.rs#L302)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> So we can pass a GlobalRef to a function that requires a T: Reflectable.
</details>
Could you just make `reflector` an intrinsic method?

---

*[components/script/dom/webidls/Promise.webidl, line 11 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12830#-KPrrhiEyePYdnX0TaS4:-KPrrhiEyePYdnX0TaS5:b-nycpan) ([raw file](https://github.com/servo/servo/blob/ee074748dad34127287fa984b4e5aee2d6358ebe/components/script/dom/webidls/Promise.webidl#L11)):*
> ```WebIDL
> // Need to escape "Promise" so it's treated as an identifier.
> interface _Promise {
> };
> ```

Is there no way to avoid doing that silly fake interface, kinda like `WindowProxy`?

---

*[components/script/dom/webidls/PromiseNativeHandler.webidl, line 13 \[r5\]](https://reviewable.io:443/reviews/servo/servo/12830#-KPrux7Qv1dqL_CCRGZk:-KPrux7Qv1dqL_CCRGZl:b9xdz9v) ([raw file](https://github.com/servo/servo/blob/05bdd4ebe22e5a4356d9bc18e28a15542433ae75/components/script/dom/webidls/PromiseNativeHandler.webidl#L13)):*
> ```WebIDL
>  Exposed=(Window,Worker)]
> interface PromiseNativeHandler {
> };
> ```

Can we avoid that? Also nit: missing new line at the end.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12830#-:-KPrvGjs4NieoDqihB2w:b-emmyac)*
<!-- Sent from Reviewable.io -->

jdm commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[components/script/dom/bindings/global.rs, line 302 \[r7\]](https://reviewable.io:443/reviews/servo/servo/12830#-KOzs7Gn2sILEkC6p9pH-r7-302:-KPs2KlMPgogr0KQ0vF9:bxnrrl) ([raw file](https://github.com/servo/servo/blob/c5ee871ce9a0bd5b56fd4e66ce992590c749935e/components/script/dom/bindings/global.rs#L302)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> Could you just make `reflector` an intrinsic method?
</details>
I'm not sure what you mean. We could split Reflectable into `Reflectable` and `MutReflectable` where the second inherits from the first and code that needs it uses it, though.

---

*[components/script/dom/webidls/Promise.webidl, line 11 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12830#-KPrrhiEyePYdnX0TaS4:-KPs2cI-YMqeoN25R2dA:buj8v51) ([raw file](https://github.com/servo/servo/blob/ee074748dad34127287fa984b4e5aee2d6358ebe/components/script/dom/webidls/Promise.webidl#L11)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> Is there no way to avoid doing that silly fake interface, kinda like `WindowProxy`?
</details>
I would rather not. We still need the rest of the items in this file, so I don't think it's a big deal.

---

*[components/script/dom/webidls/PromiseNativeHandler.webidl, line 13 \[r5\]](https://reviewable.io:443/reviews/servo/servo/12830#-KPrux7Qv1dqL_CCRGZk:-KPs2t9msy3x6g9Wh7na:b727f88) ([raw file](https://github.com/servo/servo/blob/05bdd4ebe22e5a4356d9bc18e28a15542433ae75/components/script/dom/webidls/PromiseNativeHandler.webidl#L13)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> Can we avoid that? Also nit: missing new line at the end.
</details>
We need an type that we can unwrap into safely.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12830)*
<!-- Sent from Reviewable.io -->

jdm commented:
This is blocked on https://github.com/servo/servo/pull/12954 due to breaking changes in the JS conversion APIs.

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit a726a15 with merge b380723...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2852)

jdm commented on tests/wpt/mozilla/tests/mozilla/promise.html:
```diff
@@ -0,0 +1,40 @@
+      });
+      p.then(test.unreached_func());
+      return Promise.resolve('success').then(p);
+  }, 'Native resolve callback gets argument');
```
Note to self: this test appears to be bogus? It's supposed to be testing rejection...

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit b1368a3 with merge 58a3f74...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2800)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #12980) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
Thinking more about this, I don't believe `Rc<Promise>` makes sense at all. I'm going to try making this PR use `Root<Promise>` instead, because otherwise there is a clear GC hazard when using `Promise::new` - the reflector doesn't exist anywhere else in the list of known roots.

jdm committed:
Test result updates for tests using promise_test.

jdm commented:
Trying to change Rc to Root helped remind me why I didn't do that originally - it doesn't work. The JSClass used for Promises doesn't have any special hooks for interacting with our DOM objects, so at minimum we would leak every Rust Promise object allocated. My newest changes keep the Rc, but add a permanent root that is stored inside of the Promise object, which will ensure that the "reflected" object will not be GCed before the Rust Promise object is deallocated.

jdm committed:
Ensure Promise "reflector" is not GCed before the Rust object.

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit a7068f8 with merge 0c8834f...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2875)

jdm commented:
Segfault in a test that uses promise_test. Hooray!

jdm commented:
```
  ▶ CRASH [expected TIMEOUT] /WebCryptoAPI/digest/digest.worker
  │ 
  │ Xlib:  extension "XFree86-VidModeExtension" missing on display ":0".
  │ Stack trace for thread "WebWorker for http://web-platform.test:8000/WebCryptoAPI/digest/digest.worker.js"
  │ stack backtrace:
  │    0:     0x7f780f4197cd - backtrace::backtrace::trace::h22a738598d75d447
  │    1:     0x7f780f419e92 - backtrace::capture::Backtrace::new::h1a277804460ccff5
  │    2:     0x7f780f1387ba - servo::install_crash_handler::handler::h7d01d896c8f0d15d
  │    3:     0x7f78102d5283 - AsmJSFaultHandler
  │                         at /home/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/asmjs/WasmSignalHandlers.cpp:1171
  │    4:     0x7f780d05f32f - <unknown>
  │    5:                0x0 - <unknown>
  │ Stack trace for thread "WebWorker for http://web-platform.test:8000/WebCryptoAPI/digest/digest.worker.js"
  │ stack backtrace:
  │    0:     0x7f780f4197cd - backtrace::backtrace::trace::h22a738598d75d447
  │    1:     0x7f780f419e92 - backtrace::capture::Backtrace::new::h1a277804460ccff5
  │    2:     0x7f780f1387ba - servo::install_crash_handler::handler::h7d01d896c8f0d15d
  │    3:     0x7f780caaacaf - <unknown>
  │    4:     0x7f780f13884b - servo::install_crash_handler::handler::h7d01d896c8f0d15d
  │    5:     0x7f78102d5283 - AsmJSFaultHandler
  │                         at /home/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/asmjs/WasmSignalHandlers.cpp:1171
  │    6:     0x7f780d05f32f - <unknown>
  └    7:                0x0 - <unknown>
```

jdm commented:
Also
```
  ▶ Unexpected subtest result in /html/browsers/the-window-object/window-properties.html:
  └ PASS [expected FAIL] Window replaceable attribute: self

  ▶ Unexpected subtest result in /html/browsers/the-window-object/window-properties.html:
  └ PASS [expected FAIL] Window replaceable attribute: frames

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] Window interface: attribute self

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] Window interface: attribute frames

  ▶ Unexpected subtest result in /navigation-timing/idlharness.html:
  └ PASS [expected FAIL] Window interface: attribute performance

  ▶ Unexpected subtest result in /navigation-timing/test_readwrite.html:
  └ PASS [expected FAIL] window.performance is read/write

  ▶ Unexpected subtest result in /navigation-timing/test_readwrite.html:
  └ PASS [expected FAIL] var performance is read/write
```

nox commented:
@jdm I think these unexpected passing tests are for #13043.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13043) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
Grumble. I haven't reproduced the segfault yet, even with debugmozjs and `JS_GC_ZEAL=2,1` or `JS_GC_ZEAL=1`.

jdm commented:
Oh! I don't think it's a GC-related crash at all!
```
  ▶ CRASH [expected TIMEOUT] /WebCryptoAPI/digest/digest.worker
  │ 
  │ Assertion failure: cx->runtime()->enqueuePromiseJobCallback (Must set a callback using JS_SetEnqeueuPromiseJobCallback before using Promises), at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Runtime.cpp:773
  │ Stack trace for thread "WebWorker for http://web-platform.test:8000/WebCryptoAPI/digest/digest.worker.js"
  │ stack backtrace:
  │    0:     0x7f2c481409d1 - backtrace::backtrace::libunwind::trace
  │                         at /home/jdm/servo/./.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/libunwind.rs:54
  │                          - backtrace::backtrace::trace<closure>
  │                         at /home/jdm/servo/./.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/mod.rs:70
  │    1:     0x7f2c48141580 - backtrace::capture::{{impl}}::new
  │                         at /home/jdm/servo/target/debug/build/backtrace-3bc1ff360ebb00cb/out/capture.rs:79
  │    2:     0x7f2c475bd722 - servo::install_crash_handler::handler
  │                         at /home/jdm/servo/components/servo/main.rs:70
  │    3:     0x7f2c4aaa4ed5 - AsmJSFaultHandler
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/asmjs/WasmSignalHandlers.cpp:1171
  │    4:     0x7f2c450ec33f - <unknown>
  │    5:     0x7f2c4a645d4a - 9JSRuntime17enqueuePromiseJobEP9JSContextN2JS6HandleIP10JSFunctionEENS3_IP8JSObjectE
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Runtime.cpp:772
  │    6:     0x7f2c4a6817c9 - intrinsic_EnqueuePromiseReactionJob
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/SelfHosting.cpp:1828
  │    7:     0x7f2c4a5f6434 - 2js12CallJSNativeEP9JSContextPFbS1_jPN2JS5ValueEERKNS2_8CallArgs
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/jscntxtinlines.h:232
  │    8:     0x7f2c4a5c4882 - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:453
  │    9:     0x7f2c4a5c4ba0 - InternalCall
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:498
  │   10:     0x7f2c4a5c4bca - 2js13CallFromStackEP9JSContextRKN2JS8CallArgs
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:504
  │   11:     0x7f2c4a5d1d5c - Interpret
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:2873
  │   12:     0x7f2c4a5c4500 - 2js9RunScriptEP9JSContextRNS_8RunState
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:399
  │   13:     0x7f2c4a5c4942 - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:471
  │   14:     0x7f2c4a5c4ba0 - InternalCall
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:498
  │   15:     0x7f2c4a5c4bca - 2js13CallFromStackEP9JSContextRKN2JS8CallArgs
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:504
  │   16:     0x7f2c4a5d1d5c - Interpret
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:2873
  │   17:     0x7f2c4a5c4500 - 2js9RunScriptEP9JSContextRNS_8RunState
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:399
  │   18:     0x7f2c4a5c4942 - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:471
  │   19:     0x7f2c4a5c4ba0 - InternalCall
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:498
  │   20:     0x7f2c4a5c4bca - 2js13CallFromStackEP9JSContextRKN2JS8CallArgs
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:504
  │   21:     0x7f2c4a5d1d5c - Interpret
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:2873
  │   22:     0x7f2c4a5c4500 - 2js9RunScriptEP9JSContextRNS_8RunState
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:399
  │   23:     0x7f2c4a5c4942 - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:471
  │   24:     0x7f2c4a5c4ba0 - InternalCall
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:498
  │   25:     0x7f2c4a5c4bca - 2js13CallFromStackEP9JSContextRKN2JS8CallArgs
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:504
  │   26:     0x7f2c4a5d1d5c - Interpret
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:2873
  │   27:     0x7f2c4a5c4500 - 2js9RunScriptEP9JSContextRNS_8RunState
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:399
  │   28:     0x7f2c4a5c4942 - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:471
  │   29:     0x7f2c4a5c4ba0 - InternalCall
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:498
  │   30:     0x7f2c4a5c4bca - 2js13CallFromStackEP9JSContextRKN2JS8CallArgs
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:504
  │   31:     0x7f2c4a5d1d5c - Interpret
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:2873
  │   32:     0x7f2c4a5c4500 - 2js9RunScriptEP9JSContextRNS_8RunState
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:399
  │   33:     0x7f2c4a5c4942 - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:471
  │   34:     0x7f2c4a5c4ba0 - InternalCall
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:498
  │   35:     0x7f2c4a5c4bca - 2js13CallFromStackEP9JSContextRKN2JS8CallArgs
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:504
  │   36:     0x7f2c49fc295c - DoCallFallback
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/jit/BaselineIC.cpp:5970
  │   37:     0x7f2c4700ce23 - <unknown>
  │ Stack trace for thread "WebWorker for http://web-platform.test:8000/WebCryptoAPI/digest/digest.worker.js"
  │ stack backtrace:
  │    0:     0x7f2c481409d1 - backtrace::backtrace::libunwind::trace
  │                         at /home/jdm/servo/./.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/libunwind.rs:54
  │                          - backtrace::backtrace::trace<closure>
  │                         at /home/jdm/servo/./.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/mod.rs:70
  │    1:     0x7f2c48141580 - backtrace::capture::{{impl}}::new
  │                         at /home/jdm/servo/target/debug/build/backtrace-3bc1ff360ebb00cb/out/capture.rs:79
  │    2:     0x7f2c475bd722 - servo::install_crash_handler::handler
  │                         at /home/jdm/servo/components/servo/main.rs:70
  │    3:     0x7f2c44b37d3f - <unknown>
  │    4:     0x7f2c475bd832 - servo::install_crash_handler::handler
  │                         at /home/jdm/servo/components/servo/main.rs:72
  │    5:     0x7f2c4aaa4ed5 - AsmJSFaultHandler
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/asmjs/WasmSignalHandlers.cpp:1171
  │    6:     0x7f2c450ec33f - <unknown>
  │    7:     0x7f2c4a645d4a - 9JSRuntime17enqueuePromiseJobEP9JSContextN2JS6HandleIP10JSFunctionEENS3_IP8JSObjectE
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Runtime.cpp:772
  │    8:     0x7f2c4a6817c9 - intrinsic_EnqueuePromiseReactionJob
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/SelfHosting.cpp:1828
  │    9:     0x7f2c4a5f6434 - 2js12CallJSNativeEP9JSContextPFbS1_jPN2JS5ValueEERKNS2_8CallArgs
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/jscntxtinlines.h:232
  │   10:     0x7f2c4a5c4882 - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:453
  │   11:     0x7f2c4a5c4ba0 - InternalCall
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:498
  │   12:     0x7f2c4a5c4bca - 2js13CallFromStackEP9JSContextRKN2JS8CallArgs
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:504
  │   13:     0x7f2c4a5d1d5c - Interpret
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:2873
  │   14:     0x7f2c4a5c4500 - 2js9RunScriptEP9JSContextRNS_8RunState
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:399
  │   15:     0x7f2c4a5c4942 - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:471
  │   16:     0x7f2c4a5c4ba0 - InternalCall
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:498
  │   17:     0x7f2c4a5c4bca - 2js13CallFromStackEP9JSContextRKN2JS8CallArgs
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:504
  │   18:     0x7f2c4a5d1d5c - Interpret
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:2873
  │   19:     0x7f2c4a5c4500 - 2js9RunScriptEP9JSContextRNS_8RunState
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:399
  │   20:     0x7f2c4a5c4942 - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:471
  │   21:     0x7f2c4a5c4ba0 - InternalCall
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:498
  │   22:     0x7f2c4a5c4bca - 2js13CallFromStackEP9JSContextRKN2JS8CallArgs
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:504
  │   23:     0x7f2c4a5d1d5c - Interpret
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:2873
  │   24:     0x7f2c4a5c4500 - 2js9RunScriptEP9JSContextRNS_8RunState
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:399
  │   25:     0x7f2c4a5c4942 - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:471
  │   26:     0x7f2c4a5c4ba0 - InternalCall
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:498
  │   27:     0x7f2c4a5c4bca - 2js13CallFromStackEP9JSContextRKN2JS8CallArgs
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:504
  │   28:     0x7f2c4a5d1d5c - Interpret
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:2873
  │   29:     0x7f2c4a5c4500 - 2js9RunScriptEP9JSContextRNS_8RunState
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:399
  │   30:     0x7f2c4a5c4942 - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:471
  │   31:     0x7f2c4a5c4ba0 - InternalCall
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:498
  │   32:     0x7f2c4a5c4bca - 2js13CallFromStackEP9JSContextRKN2JS8CallArgs
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:504
  │   33:     0x7f2c4a5d1d5c - Interpret
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:2873
  │   34:     0x7f2c4a5c4500 - 2js9RunScriptEP9JSContextRNS_8RunState
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:399
  │   35:     0x7f2c4a5c4942 - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:471
  │   36:     0x7f2c4a5c4ba0 - InternalCall
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:498
  │   37:     0x7f2c4a5c4bca - 2js13CallFromStackEP9JSContextRKN2JS8CallArgs
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/vm/Interpreter.cpp:504
  │   38:     0x7f2c49fc295c - DoCallFallback
  │                         at /home/jdm/servo/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/master/mozjs/js/src/jit/BaselineIC.cpp:5970
  └   39:     0x7f2c4700ce23 - <unknown>

Ran 1 tests finished in 5.0 seconds.
  • 0 ran as expected. 0 tests skipped.
  • 1 tests crashed unexpectedly
```

jdm committed:
Support promises in workers.

jdm committed:
Create meaningful tests for native promise handlers.

jdm commented:
Still need to rebase against master, but the known issues are now resolved. Further review would be appreciated.

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 42f9a5c with merge 5b1a7cd...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/3032)

highfive commented:
▶ OK [expected ERROR] /WebCryptoAPI/digest/test_digest.html
    
      ▶ OK [expected ERROR] /fetch/api/request/request-cache.html
    
      ▶ OK [expected ERROR] /fetch/api/request/request-consume-empty.html
    
      ▶ OK [expected ERROR] /fetch/api/request/request-consume.html
    
      ▶ OK [expected ERROR] /fetch/api/request/request-init-002.html
    
      ▶ OK [expected ERROR] /fetch/api/response/response-cancel-stream.html
    
      ▶ OK [expected ERROR] /fetch/api/response/response-consume-empty.html
    
      ▶ OK [expected ERROR] /fetch/api/response/response-consume-stream.html
    
      ▶ OK [expected ERROR] /fetch/api/response/response-consume.html
    
      ▶ OK [expected ERROR] /fetch/api/response/response-stream-disturbed-2.html
    
      ▶ OK [expected ERROR] /fetch/api/response/response-init-002.html
    
      ▶ OK [expected ERROR] /fetch/api/response/response-stream-disturbed-1.html
    
      ▶ OK [expected ERROR] /fetch/api/response/response-stream-disturbed-3.html
    
      ▶ OK [expected ERROR] /fetch/api/response/response-stream-disturbed-4.html
    
      ▶ OK [expected ERROR] /fetch/api/response/response-stream-disturbed-5.html

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13174) made this pull request unmergeable. Please resolve the merge conflicts.

jdm committed:
Add a simple API to reject promises with DOM error values.

Ms2ger commented:
Reviewed 1 of 2 files at r10, 1 of 4 files at r12, 1 of 6 files at r13, 1 of 5 files at r14, 2 of 6 files at r15, 6 of 12 files at r16, 16 of 17 files at r17, 1 of 8 files at r19, 2 of 5 files at r20.
Review status: 29 of 40 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[components/script/script_runtime.rs, line 106 \[r20\]](https://reviewable.io:443/reviews/servo/servo/12830#-KR9AkL4cLHUlPhI6urM:-KR9AkL4cLHUlPhI6urN:b7pjcxs) ([raw file](https://github.com/servo/servo/blob/06fda958380ce34b758fafb2132e31ad5a5bcaf3/components/script/script_runtime.rs#L106)):*
> ```Rust
> 
> #[derive(JSTraceable, HeapSizeOf)]
> pub struct EnqueuedPromiseCallback {
> ```

Docs for everything you added to this file, please.

---

*[components/script/script_runtime.rs, line 146 \[r20\]](https://reviewable.io:443/reviews/servo/servo/12830#-KR9BMHN3LlmmSb2H93b:-KR9BMHN3LlmmSb2H93c:b-5aiqu6) ([raw file](https://github.com/servo/servo/blob/06fda958380ce34b758fafb2132e31ad5a5bcaf3/components/script/script_runtime.rs#L146)):*
> ```Rust
>         for job in &*self.flushing_job_queue.borrow() {
>             if let Some(target) = target_provider(job.pipeline) {
>                 let _ = job.callback.Call_(&target.r(), ExceptionHandling::Report);
> ```

Holding the borrow on `self.flushing_job_queue` may be problematic if this can end up calling back into `enqueue` or itself. If that can't happen, please make a note (and perhaps just borrow `self.flushing_job_queue` for the entire function).

---

*[components/script/dom/promise.rs, line 240 \[r20\]](https://reviewable.io:443/reviews/servo/servo/12830#-KR9EC7_JJmyn3e4DJs9:-KR9EC7_JJmyn3e4DJsA:b-rll82e) ([raw file](https://github.com/servo/servo/blob/06fda958380ce34b758fafb2132e31ad5a5bcaf3/components/script/dom/promise.rs#L240)):*
> ```Rust
>         Ok(h) => h,
>         Err(_) => {
>             //TODO throw an exception
> ```

Please fix.

---

*[components/script/dom/promisenativehandler.rs, line 15 \[r20\]](https://reviewable.io:443/reviews/servo/servo/12830#-KR9EW_UQoBdVTtVMN5A:-KR9EW_UQoBdVTtVMN5B:bh2s1yj) ([raw file](https://github.com/servo/servo/blob/06fda958380ce34b758fafb2132e31ad5a5bcaf3/components/script/dom/promisenativehandler.rs#L15)):*
> ```Rust
> pub type CallbackType = Box<Callback>;
> 
> impl JSTraceable for Box<Callback> {
> ```

Fix the `impl<T: JSTraceable> JSTraceable for Box<T>` in `trace.rs` to allow `?Sized`?

---

*[components/script/dom/bindings/error.rs, line 273 \[r20\]](https://reviewable.io:443/reviews/servo/servo/12830#-KR9F9WRBenzdumkWpje:-KR9F9WRBenzdumkWpjf:bb5tt72) ([raw file](https://github.com/servo/servo/blob/06fda958380ce34b758fafb2132e31ad5a5bcaf3/components/script/dom/bindings/error.rs#L273)):*
> ```Rust
> impl Error {
>     /// Convert this error value to a JS value, consuming it in the process.
>     pub unsafe fn to_jsval(self, cx: *mut JSContext, global: GlobalRef, rval: MutableHandleValue) {
> ```

Needs a better name, and better docs for the case where an exception is already pending.

---

*[components/script/dom/bindings/js.rs, line 275 \[r20\]](https://reviewable.io:443/reviews/servo/servo/12830#-KR9Feqlnv59ed2pvW0q:-KR9Feqlnv59ed2pvW0r:b-s1m7z9) ([raw file](https://github.com/servo/servo/blob/06fda958380ce34b758fafb2132e31ad5a5bcaf3/components/script/dom/bindings/js.rs#L275)):*
> ```Rust
> 
>     /// Get the underlying unsafe pointer to the contained value.
>     pub fn get_unsafe(&self) -> *mut JSVal {
> ```

I'd probably mark this unsafe.

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 809 \[r20\]](https://reviewable.io:443/reviews/servo/servo/12830#-KR9FoJJ_Ezg1oiJbPBb:-KR9FoJJ_Ezg1oiJbPBc:b6z1q2u) ([raw file](https://github.com/servo/servo/blob/06fda958380ce34b758fafb2132e31ad5a5bcaf3/components/script/dom/bindings/codegen/CodegenRust.py#L809)):*
> ```Python
>             templateBody = fill(
>                 """
>                 { // Scope for our GlobalObject, ErrorResult, JSAutoCompartment,
> ```

There's only a `JSAutoCompartment` here.

---

*[components/script/dom/bindings/codegen/Configuration.py, line 198 \[r20\]](https://reviewable.io:443/reviews/servo/servo/12830#-KR9G2vLGY-Zi9qrX9AO:-KR9G2vLGY-Zi9qrX9AP:b-vucj05) ([raw file](https://github.com/servo/servo/blob/06fda958380ce34b758fafb2132e31ad5a5bcaf3/components/script/dom/bindings/codegen/Configuration.py#L198)):*
> ```Python
>         spiderMonkeyInterface = desc.get('spiderMonkeyInterface', False)
> 
>         # Callback types do not use JS smart pointers, so we should not use the
> ```

"Callback and SpiderMonkey types…"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12830)*
<!-- Sent from Reviewable.io -->

jdm committed:
Support an equivalent of Trusted<T> for Rc<Promise> objects named TrustedPromise.

jdm commented:
I added a commit to enable using Promises in async native code; haven't addressed any other review comments yet.

dati91 commented:
Hi!
The [Promise.all()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) function will be added as well? Because we are planning on using that in our tests. :)

jdm commented:
@dati91 Do you mean from native code? It already is supported in JS.

dati91 commented:
Oh, You are right! Great and sorry for bothering you, I should have checked first.

jdm commented:
@dati91 Do you mean from native code? It already is supported in JS.

jdm commented:
@Ms2ger I've addressed the review comments in the latest commit. There's one other new commit for enabling using Promise values in asynchronous operations that will need reviewing, too.

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 8f9b9d2 with merge 9ca51a4...

Ms2ger commented:
Rebased

Ms2ger commented:
Still want to have another look at the refcounting; will do that later today.

---

Reviewed 1 of 10 files at r1, 1 of 5 files at r20, 5 of 30 files at r21, 1 of 8 files at r24, 1 of 1 files at r26, 3 of 7 files at r27, 7 of 8 files at r28, 2 of 2 files at r29, 108 of 108 files at r30.
Review status: 132 of 133 files reviewed at latest revision, 20 unresolved discussions, some commit checks broke.

---

*[components/script/script_runtime.rs, line 146 at r20](https://reviewable.io:443/reviews/servo/servo/12830#-KR9BMHN3LlmmSb2H93b:-KSGYdFt0a-B5wYycPak:b5h3viw) ([raw file](https://github.com/servo/servo/blob/06fda958380ce34b758fafb2132e31ad5a5bcaf3/components/script/script_runtime.rs#L146)):*
<details><summary><i>Previously, Ms2ger wrote…</i></summary>
> Holding the borrow on `self.flushing_job_queue` may be problematic if this can end up calling back into `enqueue` or itself. If that can't happen, please make a note (and perhaps just borrow `self.flushing_job_queue` for the entire function).
</details>
So there really is no point to putting `flushing_job_queue` in a field rather than a local mutable variable, is there? Except tracing, maybe...

---

*[components/script/script_runtime.rs, line 105 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGVjaAiLWDdqodEAtz:-KSGVjaAiLWDdqodEAu-:bu09e71) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/script_runtime.rs#L105)):*
> ```Rust
> }
> 
> /// A promise callback scheduled to run during the next microtask checkpoint.
> ```

Do we support microtasks for real? If not, do we have an issue number that could improve this comment?

---

*[components/script/script_runtime.rs, line 122 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGWKW51_hl_FRL3VCd:-KSGWKW51_hl_FRL3VCe:b-aiiibg) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/script_runtime.rs#L122)):*
> ```Rust
>     /// The list of enqueued promise callbacks that will be invoked at the next microtask checkpoint.
>     promise_job_queue: DOMRefCell<Vec<EnqueuedPromiseCallback>>,
>     /// True if there is an outstanding runnable responsible for evaluating the promise job queue.
> ```

I'm not quite clear on why this is necessary; would be nice to explain.

---

*[components/script/script_runtime.rs, line 176 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGZ27d6T0D7sWG532w:-KSGZ27d6T0D7sWG532x:b-n786ob) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/script_runtime.rs#L176)):*
> ```Rust
>                                  _allocation_site: HandleObject,
>                                  _data: *mut c_void) -> bool {
>     let global = global_root_from_object(job.get());
> ```

Do we need panic protection?

---

*[components/script/dom/promise.rs, line 28 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGaEULFnTC3QtGdLgP:-KSGaEULFnTC3QtGdLgQ:b-9n6h1w) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/promise.rs#L28)):*
> ```Rust
>     reflector: Reflector,
>     #[ignore_heap_size_of = "SM handles JS values"]
>     root: MutHeapJSVal,
> ```

Terrible name :)

---

*[components/script/dom/promise.rs, line 31 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGa7oKwGsEiVJfo_YX:-KSGa7oL1vxWA_bvNFc-:b-jx4qsu) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/promise.rs#L31)):*
> ```Rust
> }
> 
> trait PromiseHelper {
> ```

Explain the purpose of this trait, please. Is it so only rc-wrapped promises can be initialized?

---

*[components/script/dom/promise.rs, line 125 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGcHV7DMvmggzBk0Nf:-KSGcHV7DMvmggzBk0Ng:b-wcirw2) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/promise.rs#L125)):*
> ```Rust
> 
>     #[allow(unsafe_code)]
>     pub fn maybe_resolve_native<T>(&self, cx: *mut JSContext, val: &T) where T: ToJSValConvertible {
> ```

What does `maybe_` mean for all those?

---

*[components/script/dom/promisenativehandler.rs, line 13 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGfpr0PC6ABaSfwtCp:-KSGfpr0PC6ABaSfwtCq:b4r27zk) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/promisenativehandler.rs#L13)):*
> ```Rust
> use js::jsapi::{JSContext, HandleValue};
> 
> pub type CallbackType = Box<Callback>;
> ```

Not convinced this typedef is useful.

---

*[components/script/dom/promisenativehandler.rs, line 29 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGfid-yOW63XB6j1sw:-KSGfid0I6TiDKwfJO6C:b-2w61bb) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/promisenativehandler.rs#L29)):*
> ```Rust
>     pub fn new(global: GlobalRef,
>            resolve: Option<CallbackType>,
>            reject: Option<CallbackType>) -> Root<PromiseNativeHandler> {
> ```

Indentation; also move the return type to the line below.

---

*[components/script/dom/promisenativehandler.rs, line 38 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGg-O5RMY9OO1r2M6N:-KSGg-O5RMY9OO1r2M6O:b-odu7sh) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/promisenativehandler.rs#L38)):*
> ```Rust
> 
>     fn callback(callback: &Option<CallbackType>, cx: *mut JSContext, v: HandleValue) {
>         if let &Some(ref callback) = callback {
> ```

Nit: `if let Some(ref callback) = *callback {}`

---

*[components/script/dom/testbinding.rs, line 695 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGfR6knneKs2CcIGQk:-KSGfR6ly1HkEKWBIYaV:bowepai) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/testbinding.rs#L695)):*
> ```Rust
>         let global = self.global();
>         let handler = PromiseNativeHandler::new(global.r(),
>                                                 resolve.map(|r| SimpleHandler::new(r)),
> ```

eta-reduce, please.

---

*[components/script/dom/workerglobalscope.rs, line 242 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGeKfsT2KwFa9tTtv9:-KSGeKftHHC533a2vIqm:btrqn5i) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/workerglobalscope.rs#L242)):*
> ```Rust
> 
>     pub fn flush_promise_jobs(&self) {
>         let _ = self.script_chan().send(CommonScriptMsg::RunnableMsg(
> ```

I'd rather unwrap the result.

---

*[components/script/dom/bindings/error.rs, line 273 at r20](https://reviewable.io:443/reviews/servo/servo/12830#-KR9F9WRBenzdumkWpje:-KSGEjObtL_sFgwwu4Ok:bgfyauc) ([raw file](https://github.com/servo/servo/blob/06fda958380ce34b758fafb2132e31ad5a5bcaf3/components/script/dom/bindings/error.rs#L273)):*
<details><summary><i>Previously, Ms2ger wrote…</i></summary>
> Needs a better name, and better docs for the case where an exception is already pending.
</details>
Let's just assert that no exception is pending; this code is not called in a situation where there can be a pending exception anyway.

---

*[components/script/dom/bindings/global.rs, line 296 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGdOpL8-DMBtz9sr3P:-KSGdOpL8-DMBtz9sr3Q:b9ak3fb) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/bindings/global.rs#L296)):*
> ```Rust
>     pub fn enqueue_promise_job(&self, job: EnqueuedPromiseCallback) {
>         match *self {
>             GlobalRef::Window(ref _window) => ScriptThread::enqueue_promise_job(job, *self),
> ```

`GlobalRef::Window(_)` is fine too

---

*[components/script/dom/bindings/refcounted.rs, line 200 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGhR0CN-xatpugNwF9:-KSGhR0CN-xatpugNwFA:befnem6) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/bindings/refcounted.rs#L200)):*
> ```Rust
>     fn addref_promise(&self, promise: Rc<Promise>) {
>         let mut table = self.promise_table.borrow_mut();
>         match table.entry(&*promise) {
> ```

I tend to prefer `table.entry().or_insert(vec![]).push(promise)`

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 661 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGgWqflv9teoR6cvK_:-KSGgWqflv9teoR6cvKa:b-t3s9sp) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/bindings/codegen/CodegenRust.py#L661)):*
> ```Python
>     # they really should be!
>     if exceptionCode is None:
>         exceptionCode = "return false;\n"
> ```

Revert, please.

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 682 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGgS92wTm3lOTfzUAX:-KSGgS92wTm3lOTfzUAY:b-j4ywbd) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/bindings/codegen/CodegenRust.py#L682)):*
> ```Python
>     # wrong type of value.
>     def onFailureNotAnObject(failureCode):
>         return CGGeneric(failureCode or
> ```

Revert this unrelated change, please.

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 824 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGh1XzgBPbuKLP1p2M:-KSGh1Y-S-1JKtP4FjCl:bq0ug36) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/bindings/codegen/CodegenRust.py#L824)):*
> ```Python
> 
>                   rooted!(in(cx) let globalObj = CurrentGlobalOrNull(cx));
>                   let _ac = JSAutoCompartment::new(cx, globalObj.handle().get());
> ```

Is this even necessary? You're entering the compartment of the global that's the global of the compartment the cx is in.

---

*[tests/wpt/mozilla/tests/mozilla/promise.html, line 61 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGUm9hRWmR90DWzBt3:-KSGUm9hRWmR90DWzBt4:b-wedw0m) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/tests/wpt/mozilla/tests/mozilla/promise.html#L61)):*
> ```HTML
>       var timeoutFired = false;
>       t.resolvePromiseDelayed(p, 'success', 100);
>       setTimeout(function() { timeoutFired = true }, 100);
> ```

This looks like an intermittent failure waiting to happen.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12830#-:-KSGhlAWNv6IEkjpEilP:b-yi1aii)*
<!-- Sent from Reviewable.io -->

jdm commented:
Review status: 132 of 133 files reviewed at latest revision, 21 unresolved discussions, some commit checks broke.

---

*[components/script/script_runtime.rs, line 146 at r20](https://reviewable.io:443/reviews/servo/servo/12830#-KR9BMHN3LlmmSb2H93b:-KSGutplE5YmGMErDPml:bt9iw4i) ([raw file](https://github.com/servo/servo/blob/06fda958380ce34b758fafb2132e31ad5a5bcaf3/components/script/script_runtime.rs#L146)):*
<details><summary><i>Previously, Ms2ger wrote…</i></summary>
> So there really is no point to putting `flushing_job_queue` in a field rather than a local mutable variable, is there? Except tracing, maybe...
</details>
There is a point - the callbacks that have not yet been executed will be traced by the GC if they're in a field.

---

*[components/script/script_runtime.rs, line 176 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGZ27d6T0D7sWG532w:-KSGvPDRULXQf-Cl0Sra:bnow9nc) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/script_runtime.rs#L176)):*
<details><summary><i>Previously, Ms2ger wrote…</i></summary>
> Do we need panic protection?
</details>
I don't see how the code would panic, so I don't see the point. It's up to you.

---

*[components/script/dom/promise.rs, line 31 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGa7oKwGsEiVJfo_YX:-KSGvhcdE6kbTXKeQtug:bb77ufg) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/promise.rs#L31)):*
<details><summary><i>Previously, Ms2ger wrote…</i></summary>
> Explain the purpose of this trait, please. Is it so only rc-wrapped promises can be initialized?
</details>
Yes.

---

*[components/script/dom/promise.rs, line 125 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGcHV7DMvmggzBk0Nf:-KSGwCMb1HNn72alguhB:b-q7k2nk) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/promise.rs#L125)):*
<details><summary><i>Previously, Ms2ger wrote…</i></summary>
> What does `maybe_` mean for all those?
</details>
I copied the pattern from Gecko, but it doesn't make sense to me. I'll remove it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12830)*
<!-- Sent from Reviewable.io -->

jdm commented:
Review status: 132 of 133 files reviewed at latest revision, 21 unresolved discussions, some commit checks broke.

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 661 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGgWqflv9teoR6cvK_:-KSH1en-Mov2GTIivYEG:b-25j4vz) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/bindings/codegen/CodegenRust.py#L661)):*
<details><summary><i>Previously, Ms2ger wrote…</i></summary>
> Revert, please.
</details>
I'd rather not. This improved the codegen output for promise code that I was looking at.

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 682 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGgS92wTm3lOTfzUAX:-KSH1lDiuILeLolrjYrm:b20jb35) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/bindings/codegen/CodegenRust.py#L682)):*
<details><summary><i>Previously, Ms2ger wrote…</i></summary>
> Revert this unrelated change, please.
</details>
Not unrelated - it made the output of the promise codegen easier to read.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12830)*
<!-- Sent from Reviewable.io -->

jdm commented:
Comments addressed, except for the ones I replied to explaining why they weren't addressed.

Ms2ger commented:
Reviewed 10 of 10 files at r31.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[components/script/script_runtime.rs, line 176 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGZ27d6T0D7sWG532w:-KSHGyKzi9mFD--XfJXx:b-b06573) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/script_runtime.rs#L176)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> I don't see how the code would panic, so I don't see the point. It's up to you.
</details>
`job` probably won't, but theoretically could be null... I'd feel happier if you added it.

---

*[components/script/dom/promise.rs, line 149 at r31](https://reviewable.io:443/reviews/servo/servo/12830#-KSHHKYWGJlpMHUuGt57:-KSHHKYWGJlpMHUuGt58:buketqp) ([raw file](https://github.com/servo/servo/blob/ab67de3094b75206bc06b9ffccbb5f46340ab6a1/components/script/dom/promise.rs#L149)):*
> ```Rust
>     #[allow(unrooted_must_root, unsafe_code)]
>     pub fn resolve(&self,
>                          cx: *mut JSContext,
> ```

Indent

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 661 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGgWqflv9teoR6cvK_:-KSHJ7HEjfqnUJ1EbSJT:b-n3y0um) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/bindings/codegen/CodegenRust.py#L661)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> I'd rather not. This improved the codegen output for promise code that I was looking at.
</details>
But none of the explicit`exceptionCode` arguments end on a newline. If this improved the codegen, you should add `exceptionCode += "\n"` or something after the if.

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 682 at r30](https://reviewable.io:443/reviews/servo/servo/12830#-KSGgS92wTm3lOTfzUAX:-KSHIxuW4otrohdxuJc-:by9smaj) ([raw file](https://github.com/servo/servo/blob/88414133732b3525df3b140c637c3fba2008c91a/components/script/dom/bindings/codegen/CodegenRust.py#L682)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> Not unrelated - it made the output of the promise codegen easier to read.
</details>
Hmm, this one actually makes sense.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12830)*
<!-- Sent from Reviewable.io -->

Ms2ger commented:
SQUASH!

-S-awaiting-review 

---

Reviewed 3 of 3 files at r32.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12830#-:-KSHdMcDVWtJz9rwhPY7:b-rsodvz)*
<!-- Sent from Reviewable.io -->

jdm committed:
Remove maybe_ prefix from Promise methods.

jdm commented:
@bors-servo: r=Ms2ger

bors-servo commented:
:pushpin: Commit 7b3e262 has been approved by `Ms2ger`

<!-- @bors-servo r=Ms2ger 7b3e262702fa3f8de54a000cf5c1826e1e9b6f36 -->

jdm commented:
@bors-servo: r=Ms2ger

bors-servo commented:
:pushpin: Commit e9c0606 has been approved by `Ms2ger`

<!-- @bors-servo r=Ms2ger e9c0606454ed50ef7f8aa0858aa211a146aebfcb -->

bors-servo commented:
:hourglass: Testing commit e9c0606 with merge 2b1a39c...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3052), [arm64](http://build.servo.org/builders/arm64/builds/3068), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3121), [linux-rel](http://build.servo.org/builders/linux-rel/builds/3295), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3123), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3119), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3246), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1014)
